src/org/exist/xslt/TemplatesFactory.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * eXist Open Source Native XML Database
+ * Copyright (C) 2001-2017 The eXist Project
+ * http://exist-db.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package org.exist.xslt;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.xml.transform.Templates;
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.sax.SAXTransformerFactory;
+import javax.xml.transform.sax.TemplatesHandler;
+import javax.xml.transform.sax.TransformerHandler;
+import org.exist.security.PermissionDeniedException;
+import org.exist.storage.DBBroker;
+import org.exist.xquery.Constants;
+import org.exist.xquery.value.NodeValue;
+import org.w3c.dom.Document;
+import org.xml.sax.SAXException;
+
+/**
+ * @author <a href="mailto:shabanovd@gmail.com">Dmitriy Shabanov</a>
+ */
+public class TemplatesFactory {
+
+  private final static ConcurrentMap<String, StylesheetUri> cache = new ConcurrentHashMap<>();
+
+  public static Stylesheet stylesheet(String stylesheet, String baseUri, Properties properties, boolean useCache) {
+
+    if (useCache && properties == null) {
+      return stylesheet(stylesheet, baseUri);
+    }
+
+    String uri = uri(stylesheet, baseUri);
+
+    return new StylesheetUri(uri, properties);
+  }
+
+  private static Stylesheet stylesheet(String stylesheet, String baseUri) {
+
+    String uri = uri(stylesheet, baseUri);
+
+    cache.computeIfAbsent(uri, StylesheetUri::new);
+    return cache.get(uri);
```

Comment:
nope, because `computeIfAbsent ` return `the previous value associated with the specified key`